netdata/netdata
# eBPF plugin CO-RE and monitoring

##### Summary
This PR is preparing `eBPF.plugin` for our next steps, among the changes observed we have:

1. Add charts to `Netdata monitoring` service. This will allow users to understand how `eBPF.plugin` is working. These new charts will also help to generate the reports for reviewers.
2. Remove old helpers and bring new helpers to load `eBPF` programs. These helpers are coming from the [functional tester](https://github.com/netdata/kernel-collector/pull/285). FOR REVIEWERS: If you start reviewing the [functional tester](https://github.com/netdata/kernel-collector/pull/285), you will see part of the code isolated and this will help you with the understanding of this PR. :smile: 
3. Prepare the plugin to work with specific versions of eBPF programs. Our goal with this is to reduce the size of `kernel-collector` releases.
4. This PR also organizing some headers and remove an unnecessary call to unlock `mutex`.
5. Finally it brings a BUG fix for kernels that has patch release bigger than `255`.

##### Component Name
ebpf.plugin
##### Test Plan

1. Clone the branch
2. Configure your `ebpf.d.conf` with at least one thread disabled, during development the following configuration file was used:

```conf
[ebpf programs]
    cachestat = yes
    dcstat = yes
    disk = yes
    filesystem = yes
    fd = yes
    hardirq = yes
    mount = yes
    oomkill = yes
    process = yes
    socket = yes
    softirq = yes
    sync = yes
    swap = yes
    #shm = yes # This should be loaded when it is commented 
    vfs = yes
    mdflush = no
```

3. Compile the branch using at least `CFLAGS="-DNETDATA_INTERNAL_CHECKS=1"`, because this will give additional information about the eBPF program loaded.
4. Run `netdata` and take a look at the table present in `Additional Information` section.

##### Additional Information

To generate the column `eBPF loaded`, I ran the following command:

```sh
$ grep eBPF /var/log/netdata/error.log  > FILE
``` 

To get the data  available at column `Data from netdata.ebpf_threads`, I ran the following command:

```sh
curl -k -o FILE https://localhost:19999/api/v1/data?chart=netdata.ebpf_threads
```

This PR was tested on the following environments:

| Linux Distribution | kernel version | eBPF loaded | Data from netdata.ebpf_threads | Expected result | Final Result |
|-------------------|----------------|--------------|--------------------------------|--------------|--------------|
| Arch Linux        | 5.16.1-arch1  | [load_arch_5_16.txt](https://github.com/netdata/netdata/files/7891643/load_arch_5_16.txt)  | [arch_5_16.txt](https://github.com/netdata/netdata/files/7891644/arch_5_16.txt) | Load 5.10 for process, 5.11 for fd, 5.15 for cachestat, and 5.4 for others. | Success |
| Slackware Current | 5.15.14  | [load_slackware_5_15.txt](https://github.com/netdata/netdata/files/7891599/load_slackware_5_15.txt)  | [slackware_5_15.txt](https://github.com/netdata/netdata/files/7891600/slackware_5_15.txt) | Load 5.10 for process, 5.11 for fd, 5.15 for cachestat, and 5.4 for others. | Success |
| Manjaro 21.1 | 5.10.89-1 |  [load_manjaro_5_10.txt](https://github.com/netdata/netdata/files/7891626/load_manjaro_5_10.txt) | [manjaro_5_10.txt](https://github.com/netdata/netdata/files/7891628/manjaro_5_10.txt)  | Load 5.10 for process and 5.4 for others.  | Success |
| Ubuntu 18.04 | 5.4.0-94 | [load_ubuntu_5_4.txt](https://github.com/netdata/netdata/files/7891651/load_ubuntu_5_4.txt) | [ubuntu_5_4.txt](https://github.com/netdata/netdata/files/7891650/ubuntu_5_4.txt) | Load only 5.4 binaries. | Success |
| Debian 10.10 | 4.19.171-2 | [load_debian_4_19.txt](https://github.com/netdata/netdata/files/7891635/load_debian_4_19.txt) | [debian_4_19.txt](https://github.com/netdata/netdata/files/7891637/debian_4_19.txt) |Load only 5.4 binaries. | Success |
| Rocky 8.5 | 4.18.0-348.7.1.el8_5 | [load_centos_4_18.txt](https://github.com/netdata/netdata/files/7891622/load_centos_4_18.txt) |  [centos_4_18.txt](https://github.com/netdata/netdata/files/7891620/centos_4_18.txt) | Load only 4.18 binaries. | Success |
| Ubuntu 18.04 | 4.15.0-166 | [load_ubuntu_4_15.txt](https://github.com/netdata/netdata/files/7891615/load_ubuntu_4_15.txt)  | [ubuntu_4_15.txt](https://github.com/netdata/netdata/files/7891616/ubuntu_4_15.txt) | Load only 4.16 binaries. | Success |
| Slackware Current | 4.14.259 | [load_slackware_4_14.txt](https://github.com/netdata/netdata/files/7891610/load_slackware_4_14.txt)  |  [slackware_4_14.txt](https://github.com/netdata/netdata/files/7891611/slackware_4_14.txt) | Load only 4.14 binaries | Success |
| CentOS 7.9 | 3.10.0-1160.49.1.el7 |  [load_centos_3_10.txt](https://github.com/netdata/netdata/files/7891605/load_centos_3_10.txt) | [centos_3_10.txt](https://github.com/netdata/netdata/files/7891606/centos_3_10.txt) | Load only 3.10 binaries | OOMkill closed, but others successed, it does not have relationship with this PR. |

Finally, it is expected that `cachestat` won't work 100% correct on kernel `5.16`, a fix for this will come with another PR.

Enjoy your review! :)
